assets/js/modules/analytics-4/datastore/custom-dimensions-gathering-data.js

```diff
@@ -0,0 +1,396 @@
+/**
+ * `modules/analytics-4` data store: custom dimensions gathering data store.
+ *
+ * Site Kit by Google, Copyright 2023 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * External dependencies
+ */
+import invariant from 'invariant';
+
+/**
+ * Internal dependencies
+ */
+import API from 'googlesitekit-api';
+import Data from 'googlesitekit-data';
+import { CORE_USER } from '../../../googlesitekit/datastore/user/constants';
+import { createFetchStore } from '../../../googlesitekit/data/create-fetch-store';
+import { createReducer } from '../../../googlesitekit/data/create-reducer';
+import { getDateString } from '../../../util';
+import { CUSTOM_DIMENSION_DEFINITIONS, MODULES_ANALYTICS_4 } from './constants';
+
+const { createRegistryControl, createRegistrySelector } = Data;
+
+const RECEIVE_CUSTOM_DIMENSION_GATHERING_DATA =
+	'RECEIVE_CUSTOM_DIMENSION_GATHERING_DATA';
+const RECEIVE_CUSTOM_DIMENSION_DATA_AVAILABLE_ON_LOAD =
+	'RECEIVE_CUSTOM_DIMENSION_DATA_AVAILABLE_ON_LOAD';
+const WAIT_FOR_CUSTOM_DIMENSION_DATA_AVAILABILITY_STATE =
+	'WAIT_FOR_CUSTOM_DIMENSION_DATA_AVAILABILITY_STATE';
+
+const fetchSaveCustomDimensionDataAvailableStateStore = createFetchStore( {
+	baseName: 'saveCustomDimensionDataAvailableState',
+	controlCallback: ( { customDimension } ) =>
```

Comment:
Please note I've used `customDimension` throughout to indicate the custom dimension "slug", rather than `customDimensionName` as per the IB.

This is for consistency with the existing codebase e.g.:

https://github.com/google/site-kit-wp/blob/61a101f8cbeb60040d5111009b410065a7be67b9/assets/js/modules/analytics-4/utils/custom-dimensions.js#L30